### ISSUE TITLE
okta-react LoginCallback not displaying error

### ISSUE THREAD

**Issue by sidsethupathi** [okta/okta-oidc-js#719]
## I'm submitting this issue for the package(s):

- [ ] jwt-verifier
- [ ] okta-angular
- [ ] oidc-middleware
- [x] okta-react
- [ ] okta-react-native
- [ ] okta-vue

## I'm submitting a:

- [x] Bug report  
- [ ] Feature request
- [ ] Other (Describe below)

## Current behavior

If Okta returns with an OAuthError (eg. "User is not assigned to the client application."), the LoginCallback doesn't display the message.

## Expected behavior

The LoginCallback should display the error.

## Minimal reproduction of the problem with instructions

1. Login with a user that isn't assigned to the client application
1. After a user logs in, the `LoginCallback` component does not display the error message.

After doing some digging, it appears that `authState.error` does exist at first:

![image](https://user-images.githubusercontent.com/2329283/77666441-f6fb0680-6f4e-11ea-9b91-e38f0fa5ab8e.png)

but then in a subsequent re-render, it `authState.error` becomes undefined which causes `LoginCallback` to return `null`.

![image](https://user-images.githubusercontent.com/2329283/77666622-2b6ec280-6f4f-11ea-9a71-2f01b347094f.png)

## Extra information about the use case/user story you are trying to implement

We want to show the error message from `LoginCallback` if there is an error message.

## Environment

- Package Version: 2.0.1
- Browser: Chrome 80.0.3987.149
- OS: OSX
- Node version (`node -v`): 10.19

**Comment by swiftone** [comment]
@sidsethupathi - Thanks for the report 

> but then in a subsequent re-render, it authState.error becomes undefined 

Can you clarify what is triggering the re-render?  Is this happening immediately (implied by your title) or is your code triggering some re-render?

**Comment by swiftone** [comment]
Internal ref: OKTA-286889

**Comment by sidsethupathi** [comment]
@swiftone The re-render happens immediately so I don't think it's anything in our code considering we are just rendering the `LoginCallback` component. That said, we aren't using `react-router-dom` so maybe our router is doing something unexpected? I will do some more digging to see if I can isolate the problem.

**Comment by sidsethupathi** [comment]
@swiftone I took the React quick start code and wasn't able to reproduce my exact error (where the user can authenticate but doesn't have permission to view the app).

However, using the quick start code, if you navigate directly to `/implicit/callback`, you won't see the expected `AuthSdkError: Unable to parse a token from the url` message displayed. However, if you step through in a debugger, you'll see it displayed in an intermediate render.

Maybe this is related? Here's a link to the simplified example. Just `yarn start` and then navigate directly to `http://localhost:3000/implicit/callback`.

**Comment by sidsethupathi** [comment]
Sorry, forgot to link the repo: https://github.com/sidsethupathi/okta-react-test

**Comment by alexspence** [comment]
I have been going in circles with intermittent errors after implementing okta on our site.  Turns out our issue is very similar to this and most likely caused by the same issue.

The issue is that the Security object is initiating an updateAuthState call which is async and not waiting for it to finish.  The LoginCallback handler code executes before the updateAuthState call in Security is complete and the AuthState gets overridden.

My solution is to duplicate the Security class and delay the render until the update auth state is finished for now.

**Comment by swiftone** [comment]
@alexspence - Absolutely correct.  I wish I had your comment a couple of days ago, as I just determined the same thing.  I have a PR about to come in, just waiting for some clean up on another PR to resolve before I push it.

Another workaround in the meantime rather than duplicating `<Security>` is to duplicate `<LoginCallback>` and make these changes:

```
const LoginCallback = () => {
  const { authService, authState } = useOktaAuth();
  const authStateReady = !authState.isPending;

  useEffect( () => {
    if( authStateReady ) { // wait until initial check is done so it doesn't wipe any error
      authService.handleAuthentication();
    }
  }, [authService, authStateReady]);

  if(authStateReady && authState.error) {
```

The reason I'd recommend this is:
 - `<LoginCallback>` is less complex than `<Security>`
 - This (waiting on the isPending to finish) is a pattern that should be used on ANY component that is interacting with changing the authState

However if that doesn't address your related issues, please share details.

**Comment by alexspence** [comment]
Thanks @swiftone I'll try this and let you know

**Comment by swiftone** [comment]
Said PR is now at https://github.com/okta/okta-oidc-js/pull/738

**Comment by alexspence** [comment]
This seems to have resolved the issue stated in this ticket, we are getting some errors that seem unrelated to this with oauth redirect params cookie on an iphone now.

**Comment by brettritter-okta** [comment]
This should be resolved in the new 3.0.1 release (which also contains a way to specify your own error reporter to allow easy customization of reporting)

**Comment by CristianoYL** [comment]
I'm still experiencing this issue in v3.0.4. If not assigned to the application, the user ends up in a blank page with route `/implicit/callback`. Passing a custom component as `errorComponent` to **LoginCallback** doesn't work either.

**Comment by swiftone** [comment]
@CristianoYL - thanks for the report.  I created a new issue to follow-up on this, as the cause may be different than was originally reported here: #926
